### ISSUE TITLE
Make EventGridPublisher more resilient

### ISSUE THREAD

**Issue by tomkerkhove** [arcus-azure/arcus.eventgrid#22]
**What feature would you like to have?**
Make EventGridPublisher more robust so that it can handle resilient errors.

**Is your feature request related to a problem? Please describe.**
When publishing fails it will not retry at the moment

**Describe the solution you'd like**
It would be good to use [Polly](https://github.com/App-vNext/Polly) to retry with exponential back-off and optionally a circuit breaker.

**Additional context**
The retry & circuit breaker should be configurable via the builder that is provided in #21

**Comment by tomkerkhove** [comment]
Moving @stijnmoreels his suggestion here:
> Just an idea: since the other issue #22 regarding the resiliency of the Event Grid Publisher, I seems to me that we should separate the "real" publishing from the more "wrapper"-like functionality (retryability, validation, security, ...) and not just writing everything in the EventGridPublisher.
This way, people could write their own "wrapper" for the publisher.

**Comment by tomkerkhove** [comment]
Personally, I would provide a step before creating the publisher where you can either use our default settings or configure the resiliency yourself - I'm not sure if Retry & Circuit Breaker should be configured separately given they have the same focus but I'm open for suggestions.

With that in mind, can you give some examples of how you would approach this, please?

**Comment by stijnmoreels** [comment]
For example, the `Retry`:

```csharp
EventGridPublisherBuilder
    .ForTopic("myTopic")
    .UsingAuthenticationKey("myAuthenticationKey")
    .Retry(3)
    .Build();
```

For example, the `CircuitBroker`:
```csharp
EventGridPublisherBuilder
    .ForTopic("myTopic")
    .UsingAuthenticationKey("myAuthenticationKey")
    .CircuitBroker(3, TimeSpan.FromSeconds(5))
    .Build();
```

Alternatively, one could also give an entire `Policy` to the builder if that's what clients would want.

**Comment by tomkerkhove** [comment]
I would prefer not to pass the policy as this is leaking internal implementation detail and forcing others to know how Polly works in my opinion.

How about:
```csharp
EventGridPublisherBuilder
    .ForTopic("myTopic")
    .UsingAuthenticationKey("myAuthenticationKey")
    .WithDefaultResiliency() // or WithoutResiliency()
    .Build();
```
and if you want full control:
```csharp
EventGridPublisherBuilder
    .ForTopic("myTopic")
    .UsingAuthenticationKey("myAuthenticationKey")
    .WithResiliency(retryCount: 3,  exceptionsAllowedBeforeBreakingCircuit: 2, durationOfBreak: TimeSpan.FromMinutes(1))
    .Build();
```

That way they can control how it should behave but we do all the heavy lifting for them.
I personally would provide circuit breaker by default.

We could still go with what you propose but then I would suggest a rename:
```csharp
EventGridPublisherBuilder
    .ForTopic("myTopic")
    .UsingAuthenticationKey("myAuthenticationKey")
    .WithRetry(3) // WithoutRetry()
    .WithCircuitBreaker(3, TimeSpan.FromSeconds(5)) // WithoutCircuitBreaker
    .Build();
```

But I'd like to keep it as simple as possible and prefer to use the above to stimulate resiliency and make it super easy.

Thought @samneirinck @SamVanhoutte?

**Comment by stijnmoreels** [comment]
Yeah, that's all good for me. I agree on every point.

**Comment by tomkerkhove** [comment]
Those are fair points as well.

Sounds like your approach is leaning towards @stijnmoreels's approach then similar to this:
```csharp
EventGridPublisherBuilder
    .ForTopic("myTopic")
    .UsingAuthenticationKey("myAuthenticationKey")
    .WithRetry(3) // No retry when not called
    .WithCircuitBreaker(3, TimeSpan.FromSeconds(5)) // or  WithCircuitBreaker(myPollyPolicy).  No circuit breaker when not called
    .Build();
```

Is that correct?

**Comment by stijnmoreels** [comment]
Ah, I just see that you would rather want an exponential retry policy.
Maybe the name `Retry` should be `ExponentialRetry` in that case. But we should probably not support a whole customization of the retry in the first phase.

**Comment by tomkerkhove** [comment]
That's a good point.

I suggest to go with this then:
```csharp
EventGridPublisherBuilder
    .ForTopic("myTopic")
    .UsingAuthenticationKey("myAuthenticationKey")
    .WithExponentialRetry(3) // No retry when not called
    .WithCircuitBreaker(3, TimeSpan.FromSeconds(5)) // No circuit breaker when not called
    .Build();
```

By default I would not support plugging in your own circuit/retry policy unless we really have a demand so that:
1. We don't leak internal usage
2. We don't provide features that nobody uses but yet need to support

Makes sense?

**Comment by stijnmoreels** [comment]
Yes, indeed.

So, we support three possible options?
* `WithExponentialRetry`
* `WithCircuitBreaker`
* `WithExponentialRetry` and `WithCircuitBreaker`

**Comment by tomkerkhove** [comment]
That's correct - `ForTopic` & `UsingAuthenticationKey` are mandatory, after that you can either call `WithExponentialRetry`, `WithCircuitBreaker` or `Build`.

**Comment by tomkerkhove** [comment]
This is implemented by @stijnmoreels - Thanks!

**Comment by tomkerkhove** [comment]
Re-opening as we don't have any docs on this yet.

This depends on #42, 	however, so that we don't have to sync it again.

**Comment by tomkerkhove** [comment]
Are you up for providing documentation on this @stijnmoreels?

Can walk you through the approach if you're interested.

**Comment by stijnmoreels** [comment]
Yeah, that would be cool. I'll check with my project manager; but that's just a formality.
